File: mcp/src/tools/gsc-analytics.ts
import { z } from 'zod';

// ============================================================================
// Valid Values for GSC Analytics
// ============================================================================

/**
 * Valid dimensions for Search Analytics queries
 * Max 3 dimensions allowed by GSC API
 */
export const VALID_DIMENSIONS = [
  'query',
  'page',
  'country',
  'device',
  'searchAppearance',
  'date',
] as const;

export type ValidDimension = typeof VALID_DIMENSIONS[number];

/**
 * Valid output formats
 */
export const VALID_FORMATS = ['json', 'csv', 'table', 'markdown'] as const;
export type ValidFormat = typeof VALID_FORMATS[number];

// ============================================================================
// GSC Analytics Run Tool
// ============================================================================

/**
 * Input schema for gsc_analytics_run tool
 *
 * Generates search analytics reports from Google Search Console.
 * This is the most important GSC tool for agents analyzing search performance.
 */
export const gscAnalyticsRunInputSchema = z.object({
  /** Site URL: domain property (sc-domain:example.com) or URL prefix (https://example.com/) */
  site: z.string().optional(),
  /** Path to configuration file (alternative to site) */
  config: z.string().optional(),
  /** Number of days to query (1-180, default: 30) */
  days: z.number().int().min(1).max(180).optional().default(30),
  /** Comma-separated dimensions (max 3): query, page, country, device, searchAppearance, date */
  dimensions: z.string().optional().default('query,page'),
  /** Maximum rows to return (1-25000, default: 100) */
  limit: z.number().int().min(1).max(25000).optional().default(100),
  /** Output format: json, csv, table, markdown (default: json) */
  format: z.enum(VALID_FORMATS).optional().default('json'),
  /** Preview query without making API call */
  dry_run: z.boolean().optional(),
}).refine(
  (data) => data.site || data.config,
  {
    message: 'Either site or config must be provided',
  }
);

export type GscAnalyticsRunInput = z.infer<typeof gscAnalyticsRunInputSchema>;

/**
 * Query metadata returned with analytics results
 */
export interface AnalyticsQueryMetadata {
  query_date: string;
  start_date: string;
  end_date: string;
  dimensions: string[];
  row_limit: number;
  filter_count: number;
}

/**
 * Aggregated metrics across all rows
 */
export interface AnalyticsAggregates {
  total_clicks: number;
  total_impressions: number;
  average_ctr: number;
  average_position: number;
}

/**
 * Single row of analytics data
 */
export interface AnalyticsRow {
  keys: string[];
  clicks: number;
  impressions: number;
  ctr: number;
  position: number;
}

/**
 * Quota status information
 */
export interface QuotaStatus {
  date: string;
  queries_used: number;
  queries_limit: number;
  remaining: number;
  percentage_used: number;
  status: 'healthy' | 'warning' | 'critical';
}

/**
 * Dry-run query preview
 */
export interface DryRunQuery {
  site_url: string;
  start_date: string;
  end_date: string;
  dimensions: string[];
  row_limit: number;
  data_state: string;
  filters?: Array<{
    dimension: string;
    operator: string;
    expression: string;
  }>;
}

/**
 * Analytics run output structure
 */
export interface AnalyticsRunOutput {
  success: boolean;
  operation: 'analytics';
  site?: string;
  period?: string;
  total_rows?: number;
  aggregates?: AnalyticsAggregates;
  rows?: AnalyticsRow[];
  metadata?: AnalyticsQueryMetadata;
  quota?: QuotaStatus;
  dry_run?: DryRunQuery;
  error?: string;
  validation_errors?: string[];
}

/**
 * Validate dimensions string
 */
export function validateDimensions(dimensionsStr: string): { valid: boolean; dimensions: string[]; errors: string[] } {
  const dimensions = dimensionsStr.split(',').map(d => d.trim()).filter(d => d.length > 0);
  const errors: string[] = [];

  if (dimensions.length === 0) {
    errors.push('At least one dimension is required');
  }

  if (dimensions.length > 3) {
    errors.push(`Maximum 3 dimensions allowed (GSC API limit), got ${dimensions.length}`);
  }

  const invalidDims = dimensions.filter(d => !VALID_DIMENSIONS.includes(d as ValidDimension));
  if (invalidDims.length > 0) {
    errors.push(`Invalid dimension(s): ${invalidDims.join(', ')}. Valid: ${VALID_DIMENSIONS.join(', ')}`);
  }

  return {
    valid: errors.length === 0,
    dimensions,
    errors,
  };
}

/**
 * Build CLI arguments for analytics run
 */
export function buildAnalyticsRunArgs(input: GscAnalyticsRunInput): string[] {
  const args: string[] = ['gsc', 'analytics', 'run'];

  if (input.config) {
    args.push('--config', input.config);
  } else if (input.site) {
    args.push('--site', input.site);
  }

  if (input.days !== undefined && input.days !== 30) {
    args.push('--days', input.days.toString());
  }

  if (input.dimensions !== undefined && input.dimensions !== 'query,page') {
    args.push('--dimensions', input.dimensions);
  }

  if (input.limit !== undefined && input.limit !== 100) {
    args.push('--limit', input.limit.toString());
  }

  if (input.format !== undefined && input.format !== 'json') {
    args.push('--format', input.format);
  }

  if (input.dry_run) {
    args.push('--dry-run');
  }

  return args;
}

/**
 * Parse CLI output for analytics run
 */
export function parseAnalyticsRunOutput(output: string): AnalyticsRunOutput {
  const result: AnalyticsRunOutput = {
    success: true,
    operation: 'analytics',
  };

  // Check for validation errors
  if (output.includes('Validation failed:')) {
    result.success = false;
    const errorMatch = output.match(/Validation failed:\s*(.+)/);
    result.error = errorMatch ? errorMatch[1].trim() : 'Validation failed';
    return result;
  }

  // Check for missing site/config error
  if (output.includes('Either --site or --config must be provided')) {
    result.success = false;
    result.error = 'Either --site or --config must be provided';
    return result;
  }

  // Check for client creation errors
  if (output.includes('Failed to create GSC client:')) {
    result.success = false;
    const errorMatch = output.match(/Failed to create GSC client:\s*(.+)/);
    result.error = errorMatch ? errorMatch[1].trim() : 'Failed to create GSC client';
    return result;
  }

  // Check for query execution errors
  if (output.includes('Failed to query search analytics:')) {
    result.success = false;
    const errorMatch = output.match(/Failed to query search analytics:\s*(.+)/);
    result.error = errorMatch ? errorMatch[1].trim() : 'Failed to query search analytics';
    return result;
  }

  // Check for config loading errors
  if (output.includes('Failed to load config:')) {
    result.success = false;
    const errorMatch = output.match(/Failed to load config:\s*(.+)/);
    result.error = errorMatch ? errorMatch[1].trim() : 'Failed to load config';
    return result;
  }

  // Check for missing search_console config
  if (output.includes('No search_console configuration found')) {
    result.success = false;
    result.error = 'No search_console configuration found in config file';
    return result;
  }

  // Parse dry-run output
  if (output.includes('Dry-run mode')) {
    return parseDryRunOutput(output);
  }

  // Parse JSON output (primary format for MCP)
  if (output.trim().startsWith('{')) {
    return parseJsonOutput(output);
  }

  // Parse table/markdown output
  return parseTableOutput(output);
}

/**
 * Parse dry-run preview output
 */
function parseDryRunOutput(output: string): AnalyticsRunOutput {
  const result: AnalyticsRunOutput = {
    success: true,
    operation: 'analytics',
  };

  const dryRun: DryRunQuery = {
    site_url: '',
    start_date: '',
    end_date: '',
    dimensions: [],
    row_limit: 0,
    data_state: 'final',
  };

  // Extract Site URL
  const siteMatch = output.match(/Site URL:\s*(.+)/);
  if (siteMatch) {
    dryRun.site_url = siteMatch[1].trim();
    result.site = siteMatch[1].trim();
  }

  // Extract Date Range
  const dateRangeMatch = output.match(/Date Range:\s*(\S+)\s+to\s+(\S+)/);
  if (dateRangeMatch) {
    dryRun.start_date = dateRangeMatch[1].trim();
    dryRun.end_date = dateRangeMatch[2].trim();
    result.period = `${dateRangeMatch[1].trim()} to ${dateRangeMatch[2].trim()}`;
  }

  // Extract Dimensions
  const dimensionsMatch = output.match(/Dimensions:\s*(.+)/);
  if (dimensionsMatch) {
    dryRun.dimensions = dimensionsMatch[1].trim().split(',').map(d => d.trim());
  }

  // Extract Row Limit
  const rowLimitMatch = output.match(/Row Limit:\s*(\d+)/);
  if (rowLimitMatch) {
    dryRun.row_limit = parseInt(rowLimitMatch[1], 10);
  }

  // Extract Data State
  const dataStateMatch = output.match(/Data State:\s*(\w+)/);
  if (dataStateMatch) {
    dryRun.data_state = dataStateMatch[1].trim();
  }

  // Extract Filters if present
  const filtersSection = output.match(/Filters:\s*([\s\S]*?)(?=\n\n|No API call made)/);
  if (filtersSection) {
    const filterRegex = /\d+\.\s*(\w+)\s+(\w+)\s+'([^']+)'/g;
    const filters: DryRunQuery['filters'] = [];
    let filterMatch;
    while ((filterMatch = filterRegex.exec(filtersSection[1])) !== null) {
      filters.push({
        dimension: filterMatch[1],
        operator: filterMatch[2],
        expression: filterMatch[3],
      });
    }
    if (filters.length > 0) {
      dryRun.filters = filters;
    }
  }

  result.dry_run = dryRun;
  return result;
}

/**
 * Parse JSON output from CLI
 */
function parseJsonOutput(output: string): AnalyticsRunOutput {
  const result: AnalyticsRunOutput = {
    success: true,
    operation: 'analytics',
  };

  try {
    // Find the JSON object in the output (may have preceding status messages)
    const jsonMatch = output.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      result.success = false;
      result.error = 'No JSON data found in output';
      return result;
    }

    const data = JSON.parse(jsonMatch[0]);

    result.site = data.SiteURL || data.site_url;
    result.period = data.Period || data.period;
    result.total_rows = data.TotalRows || data.total_rows || 0;

    // Parse aggregates
    const agg = data.Aggregates || data.aggregates;
    if (agg) {
      result.aggregates = {
        total_clicks: agg.TotalClicks || agg.total_clicks || 0,
        total_impressions: agg.TotalImpressions || agg.total_impressions || 0,
        average_ctr: agg.AverageCTR || agg.average_ctr || 0,
        average_position: agg.AveragePosition || agg.average_position || 0,
      };
    }

    // Parse rows
    const rows = data.Rows || data.rows;
    if (rows && Array.isArray(rows)) {
      result.rows = rows.map((row: Record<string, unknown>) => ({
        keys: (row.Keys || row.keys || []) as string[],
        clicks: (row.Clicks || row.clicks || 0) as number,
        impressions: (row.Impressions || row.impressions || 0) as number,
        ctr: (row.CTR || row.ctr || 0) as number,
        position: (row.Position || row.position || 0) as number,
      }));
    }

    // Parse metadata
    const meta = data.Metadata || data.metadata;
    if (meta) {
      result.metadata = {
        query_date: meta.QueryDate || meta.query_date || '',
        start_date: meta.StartDate || meta.start_date || '',
        end_date: meta.EndDate || meta.end_date || '',
        dimensions: meta.Dimensions || meta.dimensions || [],
        row_limit: meta.RowLimit || meta.row_limit || 0,
        filter_count: meta.FilterCount || meta.filter_count || 0,
      };
    }

  } catch (e) {
    result.success = false;
    result.error = `Failed to parse JSON output: ${e instanceof Error ? e.message : 'Unknown error'}`;
  }

  return result;
}

/**
 * Parse table/markdown output from CLI
 */
function parseTableOutput(output: string): AnalyticsRunOutput {
  const result: AnalyticsRunOutput = {
    success: true,
    operation: 'analytics',
  };

  // Check for no data
  if (output.includes('No data found for this query')) {
    result.total_rows = 0;
    result.rows = [];
    result.aggregates = {
      total_clicks: 0,
      total_impressions: 0,
      average_ctr: 0,
      average_position: 0,
    };
    return result;
  }

  // Extract site from output
  const siteMatch = output.match(/Querying search analytics for\s+(\S+)/);
  if (siteMatch) {
    result.site = siteMatch[1].trim();
  }

  // Extract date range
  const dateRangeMatch = output.match(/Date range:\s*(\S+)\s+to\s+(\S+)\s+\((\d+)\s+days\)/);
  if (dateRangeMatch) {
    result.period = `${dateRangeMatch[1]} to ${dateRangeMatch[2]}`;
  }

  // Extract summary section
  const periodMatch = output.match(/Period:\s*(.+)/);
  if (periodMatch) {
    result.period = periodMatch[1].trim();
  }

  const totalRowsMatch = output.match(/Total Rows:\s*(\d+)/);
  if (totalRowsMatch) {
    result.total_rows = parseInt(totalRowsMatch[1], 10);
  }

  const totalClicksMatch = output.match(/Total Clicks:\s*(\d+)/);
  const totalImpressionsMatch = output.match(/Total Impressions:\s*(\d+)/);
  const avgCtrMatch = output.match(/Average CTR:\s*([\d.]+)%/);
  const avgPositionMatch = output.match(/Avg(?:erage)? Position:\s*([\d.]+)/);

  if (totalClicksMatch || totalImpressionsMatch) {
    result.aggregates = {
      total_clicks: totalClicksMatch ? parseInt(totalClicksMatch[1], 10) : 0,
      total_impressions: totalImpressionsMatch ? parseInt(totalImpressionsMatch[1], 10) : 0,
      average_ctr: avgCtrMatch ? parseFloat(avgCtrMatch[1]) / 100 : 0,
      average_position: avgPositionMatch ? parseFloat(avgPositionMatch[1]) : 0,
    };
  }

  // Parse quota status if present
  const quotaSection = output.match(/Daily Quota Status[\s\S]*?Quota usage (\w+)/);
  if (quotaSection) {
    const quotaDateMatch = output.match(/Date:\s*(\S+)/);
    const queriesUsedMatch = output.match(/Queries Used:\s*(\d+)\s*\/\s*(\d+)\s*\(([\d.]+)%\)/);
    const remainingMatch = output.match(/Remaining:\s*(\d+)/);

    if (queriesUsedMatch) {
      const usedPercent = parseFloat(queriesUsedMatch[3]);
      let status: 'healthy' | 'warning' | 'critical' = 'healthy';
      if (usedPercent >= 95) {
        status = 'critical';
      } else if (usedPercent >= 75) {
        status = 'warning';
      }

      result.quota = {
        date: quotaDateMatch ? quotaDateMatch[1] : '',
        queries_used: parseInt(queriesUsedMatch[1], 10),
        queries_limit: parseInt(queriesUsedMatch[2], 10),
        remaining: remainingMatch ? parseInt(remainingMatch[1], 10) : 0,
        percentage_used: usedPercent,
        status,
      };
    }
  }

  return result;
}

/**
 * MCP Tool definition for gsc_analytics_run
 *
 * This is the most important GSC tool - agents will use it frequently
 * to analyze search performance, track rankings, and identify content opportunities.
 */
export const gscAnalyticsRunTool = {
  name: 'gsc_analytics_run',
  description: 'Generate search analytics reports from Google Search Console. Query search performance data including top queries, landing pages, CTR, positions, and more. This is the primary tool for SEO analysis and search performance monitoring.',
  inputSchema: {
    type: 'object',
    properties: {
      site: {
        type: 'string',
        description: 'Site URL: domain property (sc-domain:example.com) or URL prefix (https://example.com/). Required unless config is provided.',
      },
      config: {
        type: 'string',
        description: 'Path to configuration file with search_console settings. Alternative to site parameter.',
      },
      days: {
        type: 'number',
        description: 'Number of days to query (1-180). Default: 30. Data is typically 2-3 days behind.',
        default: 30,
        minimum: 1,
        maximum: 180,
      },
      dimensions: {
        type: 'string',
        description: 'Comma-separated dimensions (max 3). Valid: query, page, country, device, searchAppearance, date. Default: query,page.',
        default: 'query,page',
      },
      limit: {
        type: 'number',
        description: 'Maximum rows to return (1-25000). Default: 100.',
        default: 100,
        minimum: 1,
        maximum: 25000,
      },
      format: {
        type: 'string',
        enum: ['json', 'csv', 'table', 'markdown'],
        description: 'Output format. Use json for structured data processing. Default: json.',
        default: 'json',
      },
      dry_run: {
        type: 'boolean',
        description: 'Preview query parameters without making the API call. Useful for validating configuration.',
        default: false,
      },
    },
    oneOf: [
      { required: ['site'] },
      { required: ['config'] },
    ],
  },
};

// ============================================================================
// Exports
// ============================================================================

export const gscAnalyticsTools = [
  gscAnalyticsRunTool,
] as const;

File: cmd/gsc_analytics.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/fatih/color"
	"github.com/spf13/cobra"
	"golang.org/x/text/cases"
	"golang.org/x/text/language"

	"github.com/garbarok/ga4-manager/internal/config"
	"github.com/garbarok/ga4-manager/internal/gsc"
	"github.com/garbarok/ga4-manager/internal/render"
)

var (
	gscAnalyticsSite       string
	gscAnalyticsConfig     string
	gscAnalyticsDays       int
	gscAnalyticsDimensions string
	gscAnalyticsFormat     string
	gscAnalyticsDryRun     bool
	gscAnalyticsRowLimit   int
)

var gscAnalyticsCmd = &cobra.Command{
	Use:   "analytics",
	Short: "Search analytics and performance reporting",
	Long: `Generate reports on search performance showing how users find your site on Google.

Query search analytics data including:
- Top search queries bringing traffic
- Top landing pages from search
- Click-through rates (CTR)
- Average positions in search results
- Geographic breakdown (by country)
- Device breakdown (desktop, mobile, tablet)

Output Formats:
  - table (default): Color-coded table view in terminal
  - json: Machine-readable JSON output for automation
  - csv: CSV format for spreadsheet analysis
  - markdown: Human-readable markdown report

Data Availability:
  - Up to 16 months of historical data
  - Data is typically 2-3 days behind
  - Final (fully processed) data is used by default

Rate Limits:
  - Shares quota with URL inspection (2,000/day)
  - 600 requests per minute per property
  - Rate limiting is automatic

Examples:
  # Generate report for last 30 days (default dimensions: query, page)
  ga4 gsc analytics run --site sc-domain:example.com --days 30

  # Report with specific dimensions
  ga4 gsc analytics run --site sc-domain:example.com --days 7 --dimensions query,page,country

  # Generate from config file (recommended)
  ga4 gsc analytics run --config configs/mysite.yaml

  # Export as CSV for Excel/Sheets
  ga4 gsc analytics run --config configs/mysite.yaml --format csv > report.csv

  # Export as JSON for automation
  ga4 gsc analytics run --config configs/mysite.yaml --format json

  # Dry-run to preview query
  ga4 gsc analytics run --config configs/mysite.yaml --dry-run

Valid Dimensions (max 3):
  - query: Search queries
  - page: Landing pages
  - country: Country codes (e.g., usa, gbr, fra)
  - device: Device types (desktop, mobile, tablet)
  - searchAppearance: How the result appeared (e.g., organic, news)
  - date: Date for trend analysis`,
}

var gscAnalyticsRunCmd = &cobra.Command{
	Use:   "run",
	Short: "Generate search analytics report",
	Long:  "Query search analytics data and generate a performance report.",
	RunE:  runGSCAnalytics,
}

func init() {
	gscCmd.AddCommand(gscAnalyticsCmd)
	gscAnalyticsCmd.AddCommand(gscAnalyticsRunCmd)

	// Site URL flag (optional if using config)
	gscAnalyticsRunCmd.Flags().StringVarP(&gscAnalyticsSite, "site", "s", "", "Site URL (sc-domain:example.com or https://example.com/)")

	// Config file flag (optional if using --site)
	gscAnalyticsRunCmd.Flags().StringVarP(&gscAnalyticsConfig, "config", "c", "", "Path to configuration file")

	// Days flag (default: 30 days)
	gscAnalyticsRunCmd.Flags().IntVarP(&gscAnalyticsDays, "days", "d", 30, "Number of days to query (1-180)")

	// Dimensions flag (default: query,page)
	gscAnalyticsRunCmd.Flags().StringVar(&gscAnalyticsDimensions, "dimensions", "query,page", "Dimensions to include (comma-separated, max 3)")

	// Row limit flag (default: 100)
	gscAnalyticsRunCmd.Flags().IntVarP(&gscAnalyticsRowLimit, "limit", "l", 100, "Maximum rows to return (1-25000)")

	// Format flag (default: table)
	gscAnalyticsRunCmd.Flags().StringVarP(&gscAnalyticsFormat, "format", "f", "table", "Output format: table, json, csv, or markdown")

	// Dry-run flag
	gscAnalyticsRunCmd.Flags().BoolVar(&gscAnalyticsDryRun, "dry-run", false, "Preview query without making API call")
}

func runGSCAnalytics(cmd *cobra.Command, args []string) error {
	var siteURL string
	var dimensions []string
	var days int
	var rowLimit int

	// Load from config if provided
	if gscAnalyticsConfig != "" {
		cfg, err := config.LoadConfig(gscAnalyticsConfig)
		if err != nil {
			color.Red("✗ Failed to load config: %v", err)
			return err
		}

		if cfg.SearchConsole == nil {
			color.Red("✗ No search_console configuration found in %s", gscAnalyticsConfig)
			return fmt.Errorf("missing search_console config")
		}

		siteURL = cfg.SearchConsole.SiteURL

		// Use config values if analytics config exists
		if cfg.SearchConsole.SearchAnalytics != nil {
			// Use config date range if specified
			if cfg.SearchConsole.SearchAnalytics.DateRange != nil && cfg.SearchConsole.SearchAnalytics.DateRange.Days > 0 {
				days = cfg.SearchConsole.SearchAnalytics.DateRange.Days
			} else {
				days = gscAnalyticsDays
			}

			// Use config dimensions if specified
			if len(cfg.SearchConsole.SearchAnalytics.Dimensions) > 0 {
				dimensions = cfg.SearchConsole.SearchAnalytics.Dimensions
			} else {
				dimensions = strings.Split(gscAnalyticsDimensions, ",")
			}

			// Row limit (use default if not in config)
			rowLimit = gscAnalyticsRowLimit
		} else {
			// No analytics config, use flag defaults
			days = gscAnalyticsDays
			dimensions = strings.Split(gscAnalyticsDimensions, ",")
			rowLimit = gscAnalyticsRowLimit
		}
	} else {
		// Use flags directly
		if gscAnalyticsSite == "" {
			color.Red("✗ Either --site or --config must be provided")
			return fmt.Errorf("missing site URL or config file")
		}

		siteURL = gscAnalyticsSite
		days = gscAnalyticsDays
		dimensions = strings.Split(gscAnalyticsDimensions, ",")
		rowLimit = gscAnalyticsRowLimit
	}

	// Trim whitespace from dimensions
	for i := range dimensions {
		dimensions[i] = strings.TrimSpace(dimensions[i])
	}

	// Validate inputs
	if err := gsc.ValidateAnalyticsParams(siteURL, days, dimensions, rowLimit); err != nil {
		color.Red("✗ Validation failed: %v", err)
		return err
	}

	// Build date range
	startDate, endDate := gsc.BuildDateRange(days)

	// Build query
	query := &gsc.SearchAnalyticsQuery{
		SiteURL:    siteURL,
		StartDate:  startDate,
		EndDate:    endDate,
		Dimensions: dimensions,
		RowLimit:   rowLimit,
		DataState:  "final",
	}

	// Dry-run mode
	if gscAnalyticsDryRun {
		displayAnalyticsDryRun(query)
		return nil
	}

	// Create client
	client, err := gsc.NewClient()
	if err != nil {
		color.Red("✗ Failed to create GSC client: %v", err)
		return err
	}
	defer func() { _ = client.Close() }()

	// Execute query
	color.Cyan("📊 Querying search analytics for %s...", siteURL)
	color.Cyan("📅 Date range: %s to %s (%d days)", startDate, endDate, days)
	color.Cyan("📈 Dimensions: %s", strings.Join(dimensions, ", "))
	fmt.Println()

	report, err := client.QuerySearchAnalytics(query)
	if err != nil {
		color.Red("✗ Failed to query search analytics: %v", err)
		return err
	}

	// Display results based on format
	switch gscAnalyticsFormat {
	case "json":
		displayAnalyticsJSON(report)
	case "csv":
		displayAnalyticsCSV(report)
	case "markdown":
		displayAnalyticsMarkdown(report)
	default:
		if err := displayAnalyticsTable(report); err != nil {
			return err
		}
	}

	// Display summary and quota status
	if gscAnalyticsFormat == "table" || gscAnalyticsFormat == "markdown" {
		displayAnalyticsSummary(report)
		displayAnalyticsQuotaStatus(client)
	}

	return nil
}

func displayAnalyticsDryRun(query *gsc.SearchAnalyticsQuery) {
	color.Cyan("🔍 Dry-run mode - Preview of search analytics query")
	fmt.Println()

	color.White("Site URL:     %s", query.SiteURL)
	color.White("Date Range:   %s to %s", query.StartDate, query.EndDate)
	color.White("Dimensions:   %s", strings.Join(query.Dimensions, ", "))
	color.White("Row Limit:    %d", query.RowLimit)
	color.White("Data State:   %s", query.DataState)

	if len(query.Filters) > 0 {
		fmt.Println()
		color.Yellow("Filters:")
		for i, filter := range query.Filters {
			color.Yellow("  %d. %s %s '%s'", i+1, filter.Dimension, filter.Operator, filter.Expression)
		}
	}

	fmt.Println()
	color.Blue("ℹ️  No API call made. Remove --dry-run to execute query.")
}

// analyticsColumns builds the column list from the report's dimensions plus
// the four fixed metric columns. Title-casing the dimension names matches the
// previous hand-rolled headers.
func analyticsColumns(report *gsc.SearchAnalyticsReport) []string {
	columns := make([]string, 0, len(report.Metadata.Dimensions)+4)
	for _, dim := range report.Metadata.Dimensions {
		columns = append(columns, cases.Title(language.English).String(dim))
	}
	return append(columns, "Clicks", "Impressions", "CTR", "Position")
}

// analyticsTableRow truncates long dimension values for terminal display and
// uses one-decimal precision on CTR and position.
func analyticsTableRow(row gsc.SearchAnalyticsRow) []string {
	cells := make([]string, 0, len(row.Keys)+4)
	for _, k := range row.Keys {
		if len(k) > 60 {
			cells = append(cells, k[:57]+"...")
		} else {
			cells = append(cells, k)
		}
	}
	return append(cells,
		fmt.Sprintf("%d", row.Clicks),
		fmt.Sprintf("%d", row.Impressions),
		fmt.Sprintf("%.1f%%", row.CTR*100),
		fmt.Sprintf("%.1f", row.Position),
	)
}

// analyticsCSVRow keeps dimension values verbatim and emits CTR / position at
// full precision so downstream spreadsheet/BI tools can reformat.
func analyticsCSVRow(row gsc.SearchAnalyticsRow) []string {
	cells := make([]string, 0, len(row.Keys)+4)
	cells = append(cells, row.Keys...)
	return append(cells,
		fmt.Sprintf("%d", row.Clicks),
		fmt.Sprintf("%d", row.Impressions),
		fmt.Sprintf("%.6f", row.CTR),
		fmt.Sprintf("%.2f", row.Position),
	)
}

// analyticsMarkdownRow matches the table-mode precision; pipe escaping is
// handled inside the render package.
func analyticsMarkdownRow(row gsc.SearchAnalyticsRow) []string {
	cells := make([]string, 0, len(row.Keys)+4)
	cells = append(cells, row.Keys...)
	return append(cells,
		fmt.Sprintf("%d", row.Clicks),
		fmt.Sprintf("%d", row.Impressions),
		fmt.Sprintf("%.1f%%", row.CTR*100),
		fmt.Sprintf("%.1f", row.Position),
	)
}

func displayAnalyticsTable(report *gsc.SearchAnalyticsReport) error {
	if report.TotalRows == 0 {
		color.Yellow("⚠ No data found for this query")
		return nil
	}
	return render.Render(os.Stdout, render.FormatTable, analyticsColumns(report), report.Rows, analyticsTableRow)
}

func displayAnalyticsJSON(report *gsc.SearchAnalyticsReport) {
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	_ = encoder.Encode(report)
}

func displayAnalyticsCSV(report *gsc.SearchAnalyticsReport) {
	_ = render.Render(os.Stdout, render.FormatCSV, analyticsColumns(report), report.Rows, analyticsCSVRow)
}

func displayAnalyticsMarkdown(report *gsc.SearchAnalyticsReport) {
	fmt.Println("# Search Analytics Report")
	fmt.Println()
	fmt.Printf("**Site:** %s  \n", report.SiteURL)
	fmt.Printf("**Period:** %s  \n", report.Period)
	fmt.Printf("**Dimensions:** %s  \n", strings.Join(report.Metadata.Dimensions, ", "))
	fmt.Printf("**Generated:** %s  \n", report.Metadata.QueryDate.Format("2006-01-02 15:04:05"))
	fmt.Println()

	if report.TotalRows == 0 {
		fmt.Println("*No data found for this query*")
		return
	}

	fmt.Println("## Summary")
	fmt.Println()
	fmt.Printf("- **Total Rows:** %d\n", report.TotalRows)
	fmt.Printf("- **Total Clicks:** %d\n", report.Aggregates.TotalClicks)
	fmt.Printf("- **Total Impressions:** %d\n", report.Aggregates.TotalImpressions)
	fmt.Printf("- **Average CTR:** %.2f%%\n", report.Aggregates.AverageCTR*100)
	fmt.Printf("- **Average Position:** %.1f\n", report.Aggregates.AveragePosition)
	fmt.Println()

	fmt.Println("## Results")
	fmt.Println()

	// Limit to top 50 rows for markdown readability.
	rows := report.Rows
	if len(rows) > 50 {
		rows = rows[:50]
	}
	_ = render.Render(os.Stdout, render.FormatMarkdown, analyticsColumns(report), rows, analyticsMarkdownRow)

	if report.TotalRows > 50 {
		fmt.Println()
		fmt.Printf("*Showing top 50 of %d total rows*\n", report.TotalRows)
	}
}

func displayAnalyticsSummary(report *gsc.SearchAnalyticsReport) {
	fmt.Println()
	color.Cyan("═══ Report Summary ═══")
	fmt.Printf("Period:         %s\n", report.Period)
	fmt.Printf("Total Rows:     %d\n", report.TotalRows)
	fmt.Printf("Total Clicks:   %s\n", color.GreenString("%d", report.Aggregates.TotalClicks))
	fmt.Printf("Total Impressions: %s\n", color.BlueString("%d", report.Aggregates.TotalImpressions))
	fmt.Printf("Average CTR:    %s\n", color.YellowString("%.2f%%", report.Aggregates.AverageCTR*100))
	fmt.Printf("Avg Position:   %s\n", formatPosition(report.Aggregates.AveragePosition))
	fmt.Println()
}

func displayAnalyticsQuotaStatus(client *gsc.Client) {
	used, limit, date := client.GetQuotaStatus()
	percentage := float64(used) / float64(limit) * 100
	remaining := limit - used

	color.Cyan("═══ Daily Quota Status ═══")
	fmt.Printf("Date:           %s\n", date)
	fmt.Printf("Queries Used:   %d / %d (%.1f%%)\n", used, limit, percentage)
	fmt.Printf("Remaining:      %d\n", remaining)
	fmt.Println()

	if percentage >= 95.0 {
		color.Red("🛑 Critical: Approaching daily quota limit!")
	} else if percentage >= 75.0 {
		color.Yellow("⚠️  Warning: %.0f%% of daily quota used", percentage)
	} else {
		color.Green("✓ Quota usage healthy")
	}
	fmt.Println()
}

func formatPosition(pos float64) string {
	// Color-code position (1-3 = green, 4-10 = yellow, 10+ = red)
	if pos <= 3.0 {
		return color.GreenString("%.1f", pos)
	} else if pos <= 10.0 {
		return color.YellowString("%.1f", pos)
	}
	return color.RedString("%.1f", pos)
}
